### PR TITLE
fix(eslint-plugin): [no-namespace] fix false positive for exported namespaces when allowDeclarations=true

### LINKED ISSUE #4843: [no-namespace] Adding 'export' yields false positive when allowDeclarations=true
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
        "@typescript-eslint/no-namespace": [
            "error",
            {
                allowDeclarations: true,
            },
        ],
```

This doesn't produce any warnings if `allowDeclarations` is true

```ts
declare namespace foo {
    namespace bar {
        namespace baz {}
    }
}
```

This DOES produce a warning if `allowDeclarations` is true

```ts
declare namespace foo {
    export namespace bar {
        namespace baz {}
    }
}
```

**Expected Result**

If `allowDeclarations` is true, then neither should error.

**Actual Result**

Including `export` causes the rule to complain.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.17.0` |
| `@typescript-eslint/parser`        | `5.17.0` |
| `TypeScript`                       | `4.6.3` |
| `ESLint`                           | `8.12.0` |
| `node`                             | `17.5.0` |

### PR DESCRIPTION
**Pull request by zachkirsch**
## PR Checklist

-   [x] Addresses an existing open issue: fixes #4843
-   [ ] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

Before this PR, this doesn't produce any warnings if `allowDeclarations` is true:

```ts
declare namespace foo {
    namespace bar {
        namespace baz {}
    }
}
```

This DOES produce a warning if `allowDeclarations` is true

```ts
declare namespace foo {
    export namespace bar {
        namespace baz {}
    }
}
```

After this PR, the second example also doesn't produce a warning.

### CODE REVIEW

**Review by armano2 — DISMISSED** [review]

File: packages/eslint-plugin/src/rules/no-namespace.ts [hunk 1]
```diff
@@ -46,12 +46,15 @@ export default util.createRule<Options, MessageIds>({
...
-      );
+    function isDeclaration(node: TSESTree.Node): boolean {
+      if (
+        node.type === AST_NODE_TYPES.TSModuleDeclaration &&
+        node.declare === true
+      ) {
+        return true;
+      }
+
+      return node.parent != null && isDeclaration(node.parent);
```
Comment (armano2) [inline packages/eslint-plugin/src/rules/no-namespace.ts:57]: one thing, you should always check parent of parent

```
TSModuleDeclaration {
   body: TSModuleBlock {
      body: [
         TSModuleDeclaration {
             body: TSModuleBlock {}
         }
      ]
   }
}
```

Comment (zachkirsch) [reply packages/eslint-plugin/src/rules/no-namespace.ts:57]: Thank for the quick review! Not totally following - what should I change?

File: packages/eslint-plugin/tests/rules/no-namespace.test.ts [hunk 2]
```diff
@@ -49,6 +49,16 @@ declare namespace foo {
...
+}
+      `,
+      options: [{ allowDeclarations: true }],
+    },
+    {
+      code: `
+export declare namespace foo {
+  export namespace bar {
+    namespace baz {}
+  }
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/rules/no-namespace.test.ts:61]: I think there are few more edge cases that should be tested around declared namespaces inside a non-declared one, both exported and not:

```ts
namespace Outer {
  declare namespace Declared {
    /* export */ namespace InsideDeclared {
       // and maybe more namespaces in here
    }
  }
}
```

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
I think what @armano2 meant was that more deeply nested namespaces need a bit more checking. Once you add a few more edge case tests I think you'll see.

Otherwise looks great so far!

**Comment by zachkirsch** [comment]
@JoshuaKGoldberg thanks for the reply! I added a bunch more tests to check all the cases.

> more deeply nested namespaces need a bit more checking.

I think these are handled by the recursive call at the end of `isDeclaration()`

**Review by JoshuaKGoldberg — APPROVED** [review]
Wonderful, thanks! 👍

**Pull request merged**